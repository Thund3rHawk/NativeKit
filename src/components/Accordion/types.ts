import { ReactNode } from 'react';
import { ViewStyle, TextStyle } from 'react-native';

export interface AccordionItemProps {
  title: string;
  children: ReactNode;
  isExpanded?: boolean;
  onToggle?: () => void;
  titleStyle?: TextStyle;
  containerStyle?: ViewStyle;
  contentStyle?: ViewStyle;
  iconColor?: string;
  animationDuration?: number;
}

export interface AccordionProps {
  children: ReactNode;
  allowMultiple?: boolean;
  defaultExpandedItems?: number[];
  containerStyle?: ViewStyle;
  itemSpacing?: number;
}