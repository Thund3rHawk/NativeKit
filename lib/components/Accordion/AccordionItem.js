import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { View, Text, TouchableOpacity, StyleSheet, LayoutAnimation, Platform, UIManager, } from 'react-native';
import theme from '../../theme';
if (Platform.OS === 'android') {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
        UIManager.setLayoutAnimationEnabledExperimental(true);
    }
}
const AccordionItem = ({ title, children, isExpanded = false, onToggle, titleStyle, containerStyle, contentStyle, iconColor = theme.colors.primary, animationDuration = 300, }) => {
    const handleToggle = () => {
        LayoutAnimation.configureNext({
            duration: animationDuration,
            create: {
                type: LayoutAnimation.Types.easeInEaseOut,
                property: LayoutAnimation.Properties.opacity,
            },
            update: {
                type: LayoutAnimation.Types.easeInEaseOut,
            },
        });
        onToggle === null || onToggle === void 0 ? void 0 : onToggle();
    };
    return (_jsxs(View, { style: [styles.container, containerStyle], children: [_jsxs(TouchableOpacity, { style: styles.header, onPress: handleToggle, activeOpacity: 0.7, children: [_jsx(Text, { style: [styles.title, titleStyle], children: title }), _jsx(View, { style: [styles.icon, { borderTopColor: iconColor }], children: _jsx(View, { style: [
                                styles.iconArrow,
                                {
                                    transform: [{ rotate: isExpanded ? '180deg' : '0deg' }],
                                    borderTopColor: iconColor,
                                },
                            ] }) })] }), isExpanded && (_jsx(View, { style: [styles.content, contentStyle], children: children }))] }));
};
const styles = StyleSheet.create({
    container: {
        backgroundColor: theme.colors.surface,
        borderRadius: theme.borderRadius.md,
        marginBottom: theme.spacing.sm,
        overflow: 'hidden',
        shadowColor: theme.colors.shadow,
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.1,
        shadowRadius: 3.84,
        elevation: 5,
    },
    header: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center',
        padding: theme.spacing.md,
        backgroundColor: theme.colors.background,
    },
    title: {
        fontSize: theme.typography.fontSize.lg,
        fontWeight: theme.typography.fontWeight.semibold,
        color: theme.colors.text,
        flex: 1,
        marginRight: theme.spacing.sm,
    },
    icon: {
        width: 0,
        height: 0,
    },
    iconArrow: {
        width: 0,
        height: 0,
        borderLeftWidth: 6,
        borderRightWidth: 6,
        borderTopWidth: 8,
        borderLeftColor: 'transparent',
        borderRightColor: 'transparent',
    },
    content: {
        padding: theme.spacing.md,
        paddingTop: 0,
        backgroundColor: theme.colors.background,
    },
});
export default AccordionItem;
