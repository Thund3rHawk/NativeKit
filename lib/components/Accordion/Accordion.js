import { jsx as _jsx } from "react/jsx-runtime";
import React, { useState, Children, cloneElement } from 'react';
import { View, StyleSheet } from 'react-native';
import AccordionItem from './AccordionItem';
import theme from '../../theme';
const Accordion = ({ children, allowMultiple = false, defaultExpandedItems = [], containerStyle, itemSpacing = theme.spacing.sm, }) => {
    const [expandedItems, setExpandedItems] = useState(new Set(defaultExpandedItems));
    const handleToggle = (index) => {
        setExpandedItems(prev => {
            const newSet = new Set(prev);
            if (newSet.has(index)) {
                newSet.delete(index);
            }
            else {
                if (!allowMultiple) {
                    newSet.clear();
                }
                newSet.add(index);
            }
            return newSet;
        });
    };
    return (_jsx(View, { style: [styles.container, containerStyle], children: Children.map(children, (child, index) => {
            if (React.isValidElement(child) && child.type === AccordionItem) {
                return cloneElement(child, {
                    key: index,
                    isExpanded: expandedItems.has(index),
                    onToggle: () => handleToggle(index),
                    containerStyle: [
                        child.props.containerStyle,
                        index < Children.count(children) - 1 && { marginBottom: itemSpacing },
                    ],
                });
            }
            return child;
        }) }));
};
const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
});
export default Accordion;
