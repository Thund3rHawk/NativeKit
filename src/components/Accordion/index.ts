import Accordion from './Accordion';
import AccordionItem from './AccordionItem';

export { Accordion, AccordionItem };
export type { AccordionProps, AccordionItemProps } from './types';