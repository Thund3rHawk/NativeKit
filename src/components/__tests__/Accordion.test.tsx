import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Text } from 'react-native';
import { Accordion, AccordionItem } from '../Accordion/index';

describe('Accordion', () => {
  it('renders correctly', () => {
    const { getByText } = render(
      <Accordion>
        <AccordionItem title="Test Item">
          <Text>Test Content</Text>
        </AccordionItem>
      </Accordion>
    );

    expect(getByText('Test Item')).toBeTruthy();
  });

  it('renders badge with content', () => {
    const tree = render(
      <Accordion>
        <AccordionItem title="Test Item">
          <Text>Test Content</Text>
        </AccordionItem>
      </Accordion>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('expands and collapses on press', () => {
    const { getByText, queryByText } = render(
      <Accordion>
        <AccordionItem title="Test Item">
          <Text>Test Content</Text>
        </AccordionItem>
      </Accordion>
    );

    // Initially collapsed
    expect(queryByText('Test Content')).toBeNull();

    // Press to expand
    fireEvent.press(getByText('Test Item'));
    expect(getByText('Test Content')).toBeTruthy();

    // Press to collapse
    fireEvent.press(getByText('Test Item'));
    expect(queryByText('Test Content')).toBeNull();
  });

  it('supports multiple expanded items when allowMultiple is true', () => {
    const { getByText } = render(
      <Accordion allowMultiple={true}>
        <AccordionItem title="Item 1">
          <Text>Content 1</Text>
        </AccordionItem>
        <AccordionItem title="Item 2">
          <Text>Content 2</Text>
        </AccordionItem>
      </Accordion>
    );

    // Expand both items
    fireEvent.press(getByText('Item 1'));
    fireEvent.press(getByText('Item 2'));

    // Both should be visible
    expect(getByText('Content 1')).toBeTruthy();
    expect(getByText('Content 2')).toBeTruthy();
  });

  it('only allows one expanded item when allowMultiple is false', () => {
    const { getByText, queryByText } = render(
      <Accordion allowMultiple={false}>
        <AccordionItem title="Item 1">
          <Text>Content 1</Text>
        </AccordionItem>
        <AccordionItem title="Item 2">
          <Text>Content 2</Text>
        </AccordionItem>
      </Accordion>
    );

    // Expand first item
    fireEvent.press(getByText('Item 1'));
    expect(getByText('Content 1')).toBeTruthy();

    // Expand second item - first should close
    fireEvent.press(getByText('Item 2'));
    expect(queryByText('Content 1')).toBeNull();
    expect(getByText('Content 2')).toBeTruthy();
  });

  it('respects defaultExpandedItems prop', () => {
    const { getByText } = render(
      <Accordion defaultExpandedItems={[0]}>
        <AccordionItem title="Item 1">
          <Text>Content 1</Text>
        </AccordionItem>
        <AccordionItem title="Item 2">
          <Text>Content 2</Text>
        </AccordionItem>
      </Accordion>
    );

    // First item should be expanded by default
    expect(getByText('Content 1')).toBeTruthy();
  });
});
