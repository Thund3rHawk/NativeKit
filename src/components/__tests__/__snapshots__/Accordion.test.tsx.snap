// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders badge with content 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#F8F9FA",
          "borderRadius": 8,
          "elevation": 5,
          "marginBottom": 8,
          "overflow": "hidden",
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 3.84,
        },
        [
          undefined,
          false,
        ],
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1C1C1E",
              "flex": 1,
              "fontSize": 16,
              "fontWeight": "600",
              "marginRight": 8,
            },
            undefined,
          ]
        }
      >
        Test Item
      </Text>
      <View
        style={
          [
            {
              "height": 0,
              "width": 0,
            },
            {
              "borderTopColor": "#007AFF",
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "borderLeftColor": "transparent",
                "borderLeftWidth": 6,
                "borderRightColor": "transparent",
                "borderRightWidth": 6,
                "borderTopWidth": 8,
                "height": 0,
                "width": 0,
              },
              {
                "borderTopColor": "#007AFF",
                "transform": [
                  {
                    "rotate": "0deg",
                  },
                ],
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
